_posts/2023-10-23-unidirectional-topic-operator.md

```diff
@@ -0,0 +1,257 @@
+        Fig 2. Line graph comparing BTO and UTO max reconciliation times.<br/>
+        We used a 3-nodes cluster with default configurations, running on a local Minikube instance with 10 cores and 28 GB of memory.<br/>
+        The test driver application was running on the same machine, and the time spent on the queue was not included.
+    </small></figcaption>
+</figure>
+
+You can tune the batching mechanism by setting `STRIMZI_MAX_QUEUE_SIZE` (default: 1024), `STRIMZI_MAX_QUEUE_SIZE` (default: 100), and `MAX_BATCH_LINGER_MS` (default: 100).
+If you exceed the configured max queue size, the UTO will print an error and then shutdown (Kubernetes will take care of restarting the pod).
+In that case, you can simply raise the max queues size to avoid the periodic operator restarts.
+
+### Race condition
+
+There is a known race condition happening when the operator and external applications try to change a topic configuration concurrently.
+Given the way Kafka is implemented, there is no solution to that problem, but the UTO significantly reduces the likelihood of reconciliation failures.
+
+When the external application wins, the topics will be created using the default Kafka cluster configuration.
+In this case, the BTO simply fails the reconciliation, and you need to recover manually.
+Instead, the UTO always reconciles, which results in reconfiguration only when the default configuration differs from the spec.
+That reconfiguration will either succeed, or fail.
+
+The recommendation here is to configure the Kafka cluster with `auto.create.topics.enable: false` in order to reduce the likelihood of hitting this problem.
+Unfortunately, disabling topic creation doesn't help with applications that use the Kafka Admin client to directly create required topics (i.e. Kafka Streams based applications).
+In this case, the application should be written to wait for topic existence, or use an init container to do the same.
+
+### Finalizers
+
+[Finalizers](https://kubernetes.io/docs/concepts/overview/working-with-objects/finalizers) are used by default to avoid missing topic deletion events when the UTO is not running.
+A common pitfall is that the namespace becomes stuck in a "terminating" state when you try to delete it without first removing all finalizers.
+If this happens, you can simply remove finalizers from all `KafkaTopic` resources at once with the following command.
+
+```sh
+kubectl get kt -o yaml | yq 'del(.items[].metadata.finalizers[])' | kubectl apply -f -
+```
+
+It is also possible to disable finalizers using the following configuration.
+
+```sh
+# from Kafka custom resource
+spec:
+  entityOperator:
+    userOperator: {}
+    topicOperator: {}
+    template:
+      topicOperatorContainer:
+        env:
+          - name: STRIMZI_USE_FINALIZERS
+            value: "false"
+
+# from standalone Deployment
+spec:
+  template:
+    spec:
+      containers:
+        - name: STRIMZI_USE_FINALIZERS
+          value: "false"
+```
```

Comment:
Adding the head of the resource (The `KAfka` CR with the basic metadata a the same for Deployment) will make this easier to understand to users what are you talking about.